### ISSUE TITLE
TestEngine with ID 'kotest' failed to discover tests

### ISSUE THREAD

**Issue by viluon** [kotest/kotest#3809]
**Which version of Kotest are you using**: 5.7.2

I'm experiencing a problem with kotest running out of memory. At first glance seems different from #3758.

<details open="true">
<summary>
<h3>Test</h3>
</summary>

```kotlin
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.string.shouldHaveLength
import io.kotest.property.checkAll

class MyPropertyTest : StringSpec(
    {
        "String size" {
            checkAll<String, String> { a, b ->
                a + b shouldHaveLength a.length + b.length
            }
        }
    },
)
```

</details>

<details>
<summary>
<h3>Stack trace</h3>
</summary>

```
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 6.
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:64)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy2/jdk.proxy2.$Proxy5.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: org.junit.platform.commons.JUnitException: TestEngine with ID 'kotest' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:160)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:132)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:78)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:99)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:85)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:119)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:94)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:89)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
	... 16 more
Caused by: io.kotest.engine.config.ConfigurationException: io.github.classgraph.ClassGraphException: Uncaught exception during scan
	at io.kotest.engine.config.ConfigManager$compile$2.invoke(ConfigManager.kt:29)
	at io.kotest.engine.config.ConfigManager$compile$2.invoke(ConfigManager.kt:29)
	at io.kotest.common.ResultsKt.mapError(results.kt:17)
	at io.kotest.engine.config.ConfigManager.compile-gIAlu-s(ConfigManager.kt:29)
	at io.kotest.engine.config.ConfigManager.initialize(ConfigManager.kt:19)
	at io.kotest.runner.junit.platform.KotestJunitPlatformTestEngine.discover(KotestJunitPlatformTestEngine.kt:108)
	at io.kotest.runner.junit.platform.KotestJunitPlatformTestEngine.discover(KotestJunitPlatformTestEngine.kt:39)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
	... 26 more
Caused by: io.github.classgraph.ClassGraphException: Uncaught exception during scan
	at io.github.classgraph.ClassGraph.scan(ClassGraph.java:1637)
	at io.github.classgraph.ClassGraph.scan(ClassGraph.java:1654)
	at io.github.classgraph.ClassGraph.scan(ClassGraph.java:1667)
	at io.kotest.engine.config.ApplyConfigFromAutoScanKt.applyConfigFromAutoScan(applyConfigFromAutoScan.kt:24)
	at io.kotest.engine.config.ConfigManager.compile-gIAlu-s(ConfigManager.kt:27)
	... 30 more
Caused by: java.lang.OutOfMemoryError: Java heap space
	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler$7.<init>(NestedJarHandler.java:668)
	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler.openInflaterInputStream(NestedJarHandler.java:663)
	at nonapi.io.github.classgraph.fileslice.Slice.open(Slice.java:246)
	at io.github.classgraph.ClasspathElementZip$1.open(ClasspathElementZip.java:390)
	at io.github.classgraph.ClasspathElementZip$1.openClassfile(ClasspathElementZip.java:376)
	at io.github.classgraph.Classfile.<init>(Classfile.java:2002)
	at io.github.classgraph.Scanner$ClassfileScannerWorkUnitProcessor.processWorkUnit(Scanner.java:725)
	at io.github.classgraph.Scanner$ClassfileScannerWorkUnitProcessor.processWorkUnit(Scanner.java:647)
	at nonapi.io.github.classgraph.concurrency.WorkQueue.runWorkLoop(WorkQueue.java:246)
	at nonapi.io.github.classgraph.concurrency.WorkQueue.access$000(WorkQueue.java:50)
	at nonapi.io.github.classgraph.concurrency.WorkQueue$1.call(WorkQueue.java:201)
	at nonapi.io.github.classgraph.concurrency.WorkQueue$1.call(WorkQueue.java:198)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.runWith(Thread.java:1596)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```
</details>

**Comment by viluon** [comment]
Update: seems to occur when running tests via Gradle, both when I run all tests and when I select only the Kotest class, like so:
```bash
gradle :core:test --tests "some.path.MyPropertyTest"
```
Running only the `String size` test via the IntelliJ Kotest plugin works fine. Note that other tests in the project don't use Kotest and without Kotest they work fine. This is the only Kotest class.

Still a problem in 5.8.0.

**Comment by Kantis** [comment]
Your stacktrace indicate that you run out of heap. Could you try adding `org.gradle.jvmargs=-Xmx3g -XX:MaxMetaspaceSize=756m` in `gradle.properties`?

**Comment by viluon** [comment]
I replaced the previous configuration value of `org.gradle.jvmargs="-Xmx256m"` (sorry for not reporting that, slipped my mind) with `org.gradle.jvmargs=-Xmx3g -XX:MaxMetaspaceSize=756m`. The issue persists -- seems this much memory should be more than enough for tests.

#### Other relevant configuration
In `build.gradle.kts`:
```kt
    testImplementation("io.kotest:kotest-runner-junit5:5.8.0")
    testImplementation("io.kotest:kotest-property:5.8.0")
```

Gradle: 8.5

JDK information:
```
JAVA_VERSION=21.0.1
JAVA_VENDOR=Azul Systems, Inc.
RUNTIME_NAME=OpenJDK Runtime Environment
RUNTIME_VERSION=21.0.1+12-LTS
VM_NAME=OpenJDK 64-Bit Server VM
VM_VERSION=21.0.1+12-LTS
VM_VENDOR=Azul Systems, Inc.
OS_ARCH=amd64
```

Gradle-reported OS information:
```
os.detected.name=linux
os.detected.arch=x86_64
os.detected.bitness=64
os.detected.version=6.2
os.detected.version.major=6
os.detected.version.minor=2
os.detected.release=ubuntu
os.detected.release.version=22.04
os.detected.release.like.ubuntu=true
os.detected.release.like.debian=true
os.detected.classifier=linux-x86_64
```

**Comment by Kantis** [comment]
Indeed it does. Could you add ` -XX:+HeapDumpOnOutOfMemoryError` and see if you see anything of in the dump?

**Comment by viluon** [comment]
Although the JDK I use [reportedly supports `HeapDumpOnOutOfMemoryError`](https://chriswhocodes.com/zulu_jdk21_options.html) and accepts the option (verified by specifying `-XX:HeapDumpOnOutOfMemoryError=true`, which leads to an instant crash about missing `+`), it does not seem to produce a heap dump in either the CWD or in `/tmp/dumps/`, explicitly requested with `-XX:HeapDumpPath`. I'm not sure what's going on.

**Comment by viluon** [comment]
I stopped all Gradle daemons and searched running processes while executing the Gradle test task. The process parameters contain the options from `gradle.properties` and run on the reported JDK. The heap dump options are passed too, though I'm yet to find a dump file in any directory. Everything seems to point to a space leak within Kotest or its dependencies -- the task fails in less than 2 seconds during test discovery.

I'll try to profile the heap to gain some insight into what's going on.

**Comment by viluon** [comment]
:man_facepalming: I missed this section of a global `build.gradle.kts` (the Kotest dependency is defined in a module-specific `build.gradle.kts`):
```kotlin
tasks.withType<Test> {
    useJUnitPlatform()
    maxHeapSize = "64m"
}
```

The processes in charge of test discovery apparently don't inherit settings from `gradle.properties`, which is why I wasn't getting a heap dump. Adding options like `+HeapDumpOnOutOfMemoryError` via `jvmArgs` works as expected.

I was suspicious when a manual heap dump via VisualVM showed only 64 MiB total size. When I changed the limits to 32 GB in `gradle.properties`, the manual heap dump was still only 64 MiB and the test still took about 2 seconds to fail. I then scanned the whole project for the `64m` limit and sure enough, there it was.

Apologies for taking your time.
